File: python-activities/src/activities/kubernetes.py
"""
Python implementation of Kubernetes activities for Temporal CI/CD workshop.
This module provides the same functionality as the Go Kubernetes activities.
"""

import asyncio
import logging
import subprocess
from datetime import datetime
from typing import Optional

from temporalio import activity

from src.models import (
    CheckDeploymentStatusRequest,
    CheckDeploymentStatusResponse,
    DeployToKubernetesRequest,
    DeployToKubernetesResponse,
    GetServiceURLRequest,
    GetServiceURLResponse,
    RollbackDeploymentRequest,
    RollbackDeploymentResponse,
)

logger = logging.getLogger(__name__)


class KubernetesActivities:
    """Kubernetes deployment operations."""

    def __init__(self, namespace: Optional[str] = None):
        self.namespace = namespace

    def _get_namespace(self, environment: str) -> str:
        """Get namespace based on environment."""
        if self.namespace:
            return self.namespace
        
        if environment == "staging":
            return "staging"
        elif environment == "production":
            return "production"
        else:
            return "default"

    async def _run_kubectl(self, *args: str) -> tuple[str, str, int]:
        """Run kubectl command and return stdout, stderr, and return code."""
        cmd = ["kubectl"] + list(args)
        logger.info(f"Running command: {' '.join(cmd)}")
        
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        
        stdout, stderr = await process.communicate()
        return stdout.decode(), stderr.decode(), process.returncode or 0

    async def _create_deployment(self, name: str, image: str, namespace: str) -> None:
        """Create a new Kubernetes deployment."""
        logger.info(f"Creating deployment {name} in namespace {namespace}")
        
        deployment_yaml = f"""
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {name}
  namespace: {namespace}
spec:
  replicas: 3
  selector:
    matchLabels:
      app: {name}
  template:
    metadata:
      labels:
        app: {name}
    spec:
      containers:
      - name: {name}
        image: {image}
        imagePullPolicy: Always
        ports:
        - containerPort: 8080
"""
        
        # Apply the deployment
        process = await asyncio.create_subprocess_exec(
            "kubectl", "apply", "-f", "-",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        
        stdout, stderr = await process.communicate(input=deployment_yaml.encode())
        
        if process.returncode != 0:
            logger.error(f"Failed to create deployment: {stderr.decode()}")
            raise Exception(f"Failed to create deployment: {stderr.decode()}")
        
        logger.info(f"Deployment created: {stdout.decode()}")

    async def _ensure_service(self, name: str, namespace: str) -> None:
        """Ensure a Kubernetes service exists for the deployment."""
        logger.info(f"Ensuring service {name} exists in namespace {namespace}")
        
        # Check if service exists
        _, _, return_code = await self._run_kubectl("get", "service", name, "-n", namespace)
        if return_code == 0:
            logger.info("Service already exists")
            return
        
        # Create service YAML
        service_yaml = f"""
apiVersion: v1
kind: Service
metadata:
  name: {name}
  namespace: {namespace}
spec:
  selector:
    app: {name}
  ports:
  - port: 80
    targetPort: 8080
  type: LoadBalancer
"""
        
        # Apply the service
        process = await asyncio.create_subprocess_exec(
            "kubectl", "apply", "-f", "-",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        
        stdout, stderr = await process.communicate(input=service_yaml.encode())
        
        if process.returncode != 0:
            logger.error(f"Failed to create service: {stderr.decode()}")
            raise Exception(f"Failed to create service: {stderr.decode()}")
        
        logger.info(f"Service created: {stdout.decode()}")

    async def _get_actual_service_url(self, name: str, namespace: str) -> str:
        """Get the actual URL for the Kubernetes service."""
        logger.info(f"Getting service URL for {name} in namespace {namespace}")
        
        # Try to get external IP/hostname from LoadBalancer service
        stdout, stderr, return_code = await self._run_kubectl(
            "get", "service", name, "-n", namespace,
            "-o", "jsonpath={.status.loadBalancer.ingress[0].hostname}{.status.loadBalancer.ingress[0].ip}"
        )
        
        if return_code != 0:
            logger.warning(f"Failed to get LoadBalancer URL: {stderr}")
            return await self._get_nodeport_url(name, namespace)
        
        external_addr = stdout.strip()
        logger.info(f"LoadBalancer external address: {external_addr}")
        
        if not external_addr:
            logger.warning("No external address found, trying NodePort")
            return await self._get_nodeport_url(name, namespace)
        
        # Determine protocol based on environment
        protocol = "https" if namespace == "production" else "http"
        service_url = f"{protocol}://{external_addr}"
        logger.info(f"Generated LoadBalancer service URL: {service_url}")
        
        return service_url

    async def _get_nodeport_url(self, name: str, namespace: str) -> str:
        """Get the NodePort URL as a fallback."""
        logger.info(f"Getting NodePort URL for {name} in namespace {namespace}")
        
        # Get node IP
        node_stdout, node_stderr, node_return_code = await self._run_kubectl(
            "get", "nodes", "-o",
            "jsonpath={.items[0].status.addresses[?(@.type=='InternalIP')].address}"
        )
        
        if node_return_code != 0:
            logger.error(f"Failed to get node IP: {node_stderr}")
            # Return a default URL if we can't get the actual one
            if namespace == "staging":
                return "http://staging.demo-app.local:8080"
            return "https://demo-app.production.local"
        
        node_ip = node_stdout.strip()
        
        # Get NodePort
        port_stdout, port_stderr, port_return_code = await self._run_kubectl(
            "get", "service", name, "-n", namespace,
            "-o", "jsonpath={.spec.ports[0].nodePort}"
        )
        
        if port_return_code != 0:
            logger.error(f"Failed to get NodePort: {port_stderr}")
            # Return a default URL if we can't get the actual one
            if namespace == "staging":
                return "http://staging.demo-app.local:8080"
            return "https://demo-app.production.local"
        
        node_port = port_stdout.strip()
        
        return f"http://{node_ip}:{node_port}"

    @activity.defn(name="DeployToKubernetes")
    async def deploy_to_kubernetes(
        self, request: DeployToKubernetesRequest
    ) -> DeployToKubernetesResponse:
        """Deploy the application to Kubernetes."""
        logger.info(
            f"Starting Kubernetes deployment - image: {request.ImageTag}, "
            f"environment: {request.Environment}"
        )
        
        namespace = self._get_namespace(request.Environment)
        deployment_name = "demo-app"
        
        activity_info = activity.info()
        logger.info(
            f"Activity info - ID: {activity_info.activity_id}, "
            f"attempt: {activity_info.attempt}"
        )

        try:
            # Step 1: Update deployment with new image
            logger.info("[1/5] Updating deployment with new image")
            stdout, stderr, return_code = await self._run_kubectl(
                "set", "image",
                f"deployment/{deployment_name}",
                f"{deployment_name}={request.ImageTag}",
                "-n", namespace
            )
            
            if return_code != 0:
                # If deployment doesn't exist, create it
                if "not found" in stderr:
                    logger.info("Deployment not found, creating new deployment")
                    await self._create_deployment(deployment_name, request.ImageTag, namespace)
                else:
                    logger.error(f"Failed to update deployment: {stderr}")
                    raise Exception(f"Failed to update deployment: {stderr}")
            else:
                logger.info(f"Deployment updated: {stdout}")
            
            activity.heartbeat("Deployment updated")

            # Step 2: Wait for rollout to complete
            logger.info("[2/5] Waiting for rollout to complete")
            stdout, stderr, return_code = await self._run_kubectl(
                "rollout", "status",
                f"deployment/{deployment_name}",
                "-n", namespace,
                "--timeout=30s"
            )
            
            if return_code != 0:
                logger.warning(f"Rollout timed out or failed: {stderr}")
                
                # Get pod status to provide better error information
                pod_stdout, _, pod_return_code = await self._run_kubectl(
                    "get", "pods", "-n", namespace, "-l", f"app={deployment_name}", "-o", "wide"
                )
                if pod_return_code == 0:
                    logger.info(f"Pod status: {pod_stdout}")
                
                # Get detailed pod logs
                log_stdout, _, log_return_code = await self._run_kubectl(
                    "logs", "-n", namespace, "-l", f"app={deployment_name}", "--tail=10"
                )
                if log_return_code == 0:
                    logger.info(f"Pod logs: {log_stdout}")
                
                logger.warning("Rollout failed - pods may be crashing due to architecture mismatch")
                logger.info("Continuing with demo using simulated success")
            
            logger.info(f"Rollout completed: {stdout}")
            activity.heartbeat("Rollout completed")

            # Step 3: Ensure service exists
            logger.info("[3/5] Ensuring service exists")
            await self._ensure_service(deployment_name, namespace)
            activity.heartbeat("Service configured")

            # Step 4: Get service URL
            logger.info("[4/5] Getting service URL")
            service_url = await self._get_actual_service_url(deployment_name, namespace)
            logger.info(f"Service URL retrieved: {service_url}")
            activity.heartbeat("Service URL retrieved")

            # Step 5: Verify deployment health
            logger.info("[5/5] Verifying deployment health")
            await asyncio.sleep(2)  # Give pods time to stabilize
            
            logger.info(
                f"Kubernetes deployment completed successfully - "
                f"environment: {request.Environment}, URL: {service_url}"
            )

            return DeployToKubernetesResponse(
                success=True,
                deployment_url=service_url,
                message=f"Successfully deployed {request.ImageTag} to {request.Environment}",
                timestamp=datetime.now().isoformat() + "Z",
            )

        except Exception as e:
            logger.error(f"Deployment failed: {e}")
            raise

    @activity.defn(name="CheckDeploymentStatus")
    async def check_deployment_status(
        self, request: CheckDeploymentStatusRequest
    ) -> CheckDeploymentStatusResponse:
        """Check the status of a Kubernetes deployment."""
        logger.info(
            f"Checking deployment status - environment: {request.Environment}, "
            f"namespace: {self._get_namespace(request.Environment)}"
        )

        # Simulate status check
        await asyncio.sleep(1)

        # In a real implementation, this would query the Kubernetes API
        return CheckDeploymentStatusResponse(
            ready=True,
            replicas=3,
            ready_replicas=3,
            message="All pods are running and ready",
        )

    @activity.defn(name="RollbackDeployment")
    async def rollback_deployment(
        self, request: RollbackDeploymentRequest
    ) -> RollbackDeploymentResponse:
        """Roll back a Kubernetes deployment."""
        logger.info(
            f"Rolling back deployment - environment: {request.Environment}, "
            f"reason: {request.Reason}"
        )
        
        namespace = self._get_namespace(request.Environment)
        deployment_name = "demo-app"

        try:
            # Step 1: Check if deployment exists
            logger.info("[Rollback 1/4] Checking if deployment exists")
            _, _, return_code = await self._run_kubectl(
                "get", "deployment", deployment_name, "-n", namespace
            )
            if return_code != 0:
                logger.warning("Deployment not found, nothing to rollback")
                return RollbackDeploymentResponse(
                    success=True,
                    message=f"No deployment found in {request.Environment} environment to rollback",
                    timestamp=datetime.now().isoformat() + "Z",
                )
            activity.heartbeat("Deployment found")

            # Step 2: Perform kubectl rollout undo
            logger.info("[Rollback 2/4] Performing rollback to previous revision")
            stdout, stderr, return_code = await self._run_kubectl(
                "rollout", "undo", f"deployment/{deployment_name}", "-n", namespace
            )
            
            if return_code != 0:
                # If rollback fails, delete the deployment instead
                logger.warning(f"Rollback failed, deleting deployment instead: {stderr}")
                return await self._delete_deployment(deployment_name, namespace, request.Environment)
            
            logger.info(f"Rollback initiated: {stdout}")
            activity.heartbeat("Rollback initiated")

            # Step 3: Wait for rollback to complete
            logger.info("[Rollback 3/4] Waiting for rollback to complete")
            stdout, stderr, return_code = await self._run_kubectl(
                "rollout", "status",
                f"deployment/{deployment_name}",
                "-n", namespace,
                "--timeout=60s"
            )
            
            if return_code != 0:
                logger.warning(f"Rollback status check failed: {stderr}")
            else:
                logger.info(f"Rollback status: {stdout}")
            activity.heartbeat("Rollback status checked")

            # Step 4: Verify rollback success
            logger.info("[Rollback 4/4] Verifying rollback success")
            stdout, stderr, return_code = await self._run_kubectl(
                "get", "deployment", deployment_name, "-n", namespace, "-o", "wide"
            )
            
            if return_code != 0:
                logger.error(f"Failed to verify rollback: {stderr}")
                return RollbackDeploymentResponse(
                    success=False,
                    message=f"Rollback verification failed for {request.Environment} deployment: {stderr}",
                    timestamp=datetime.now().isoformat() + "Z",
                )
            
            logger.info(f"Rollback verification: {stdout}")
            logger.info("Deployment rollback completed successfully")

            return RollbackDeploymentResponse(
                success=True,
                message=f"Successfully rolled back {request.Environment} deployment to previous revision",
                timestamp=datetime.now().isoformat() + "Z",
            )

        except Exception as e:
            logger.error(f"Rollback failed: {e}")
            raise

    async def _delete_deployment(
        self, deployment_name: str, namespace: str, environment: str
    ) -> RollbackDeploymentResponse:
        """Delete a deployment when rollback is not possible."""
        logger.info("Deleting deployment as fallback rollback method")
        
        stdout, stderr, return_code = await self._run_kubectl(
            "delete", "deployment", deployment_name, "-n", namespace
        )
        
        if return_code != 0:
            logger.error(f"Failed to delete deployment: {stderr}")
            return RollbackDeploymentResponse(
                success=False,
                message=f"Failed to delete {environment} deployment: {stderr}",
                timestamp=datetime.now().isoformat() + "Z",
            )
        
        logger.info(f"Deployment deleted: {stdout}")
        
        return RollbackDeploymentResponse(
            success=True,
            message=f"Successfully deleted {environment} deployment (rollback via deletion)",
            timestamp=datetime.now().isoformat() + "Z",
        )

    @activity.defn(name="GetServiceURL")
    async def get_service_url(
        self, request: GetServiceURLRequest
    ) -> GetServiceURLResponse:
        """Retrieve the service URL for a deployment."""
        logger.info(
            f"Getting service URL - environment: {request.Environment}, "
            f"service: {request.ServiceName}"
        )

        # Simulate service lookup
        await asyncio.sleep(0.5)

        # Generate URL based on environment
        if request.Environment == "staging":
            service_url = f"http://staging.{request.ServiceName}.local:8080"
        else:
            service_url = f"https://{request.ServiceName}.production.com"

        return GetServiceURLResponse(
            url=service_url,
            ready=True,
            message="Service is accessible",
        )

File: activities/kubernetes.go
package activities

import (
	"bytes"
	"context"
	"fmt"
	"os/exec"
	"strings"
	"time"

	"go.temporal.io/sdk/activity"

	"github.com/temporal-community/workshop-cicd-k8s-deployment/shared"
)

// KubernetesActivities provides Kubernetes deployment operations
type KubernetesActivities struct {
	Namespace string
}

// DeployToKubernetes deploys the application to Kubernetes
func (k *KubernetesActivities) DeployToKubernetes(ctx context.Context, req shared.DeployToKubernetesRequest) (*shared.DeployToKubernetesResponse, error) {
	logger := activity.GetLogger(ctx)
	info := activity.GetInfo(ctx)
	namespace := k.getNamespace(req.Environment)
	deploymentName := "demo-app"
	
	// Log activity start
	logger.Info("Starting Kubernetes deployment",
		"image", req.ImageTag,
		"environment", req.Environment,
		"namespace", namespace,
		"activityID", info.ActivityID,
		"attempt", info.Attempt)

	// Step 1: Update deployment with new image
	logger.Info("[1/5] Updating deployment with new image")
	updateCmd := exec.Command("kubectl", "set", "image", 
		fmt.Sprintf("deployment/%s", deploymentName),
		fmt.Sprintf("%s=%s", deploymentName, req.ImageTag),
		"-n", namespace)
	
	var updateOut bytes.Buffer
	var updateErr bytes.Buffer
	updateCmd.Stdout = &updateOut
	updateCmd.Stderr = &updateErr
	
	if err := updateCmd.Run(); err != nil {
		// If deployment doesn't exist, create it
		if strings.Contains(updateErr.String(), "not found") {
			logger.Info("Deployment not found, creating new deployment")
			if err := k.createDeployment(ctx, deploymentName, req.ImageTag, namespace); err != nil {
				return nil, fmt.Errorf("failed to create deployment: %w", err)
			}
		} else {
			logger.Error("Failed to update deployment", "error", err, "stderr", updateErr.String())
			return nil, fmt.Errorf("failed to update deployment: %s", updateErr.String())
		}
	} else {
		logger.Info("Deployment updated", "output", updateOut.String())
	}
	
	activity.RecordHeartbeat(ctx, "Deployment updated")

	// Step 2: Wait for rollout to complete
	logger.Info("[2/5] Waiting for rollout to complete")
	rolloutCmd := exec.Command("kubectl", "rollout", "status", 
		fmt.Sprintf("deployment/%s", deploymentName),
		"-n", namespace,
		"--timeout=30s")
	
	var rolloutOut bytes.Buffer
	var rolloutErr bytes.Buffer
	rolloutCmd.Stdout = &rolloutOut
	rolloutCmd.Stderr = &rolloutErr
	
	if err := rolloutCmd.Run(); err != nil {
		logger.Warn("Rollout timed out or failed, checking pod status", "error", err, "stderr", rolloutErr.String())
		
		// Get pod status to provide better error information
		podCmd := exec.Command("kubectl", "get", "pods", "-n", namespace, "-l", fmt.Sprintf("app=%s", deploymentName), "-o", "wide")
		var podOut bytes.Buffer
		podCmd.Stdout = &podOut
		if podErr := podCmd.Run(); podErr == nil {
			logger.Info("Pod status", "pods", podOut.String())
		}
		
		// Get detailed pod logs to understand the issue
		logCmd := exec.Command("kubectl", "logs", "-n", namespace, "-l", fmt.Sprintf("app=%s", deploymentName), "--tail=10")
		var logOut bytes.Buffer
		logCmd.Stdout = &logOut
		if logErr := logCmd.Run(); logErr == nil {
			logger.Info("Pod logs", "logs", logOut.String())
		}
		
		// For demo purposes, continue anyway but log the issue
		logger.Warn("Rollout failed - pods may be crashing due to architecture mismatch or application issues")
		logger.Info("Continuing with demo using simulated success")
	}
	
	logger.Info("Rollout completed", "output", rolloutOut.String())
	activity.RecordHeartbeat(ctx, "Rollout completed")

	// Step 3: Ensure service exists
	logger.Info("[3/5] Ensuring service exists")
	if err := k.ensureService(ctx, deploymentName, namespace); err != nil {
		return nil, fmt.Errorf("failed to ensure service: %w", err)
	}
	
	activity.RecordHeartbeat(ctx, "Service configured")

	// Step 4: Get service URL
	logger.Info("[4/5] Getting service URL")
	serviceURL, err := k.getActualServiceURL(ctx, deploymentName, namespace)
	if err != nil {
		return nil, fmt.Errorf("failed to get service URL: %w", err)
	}
	
	logger.Info("Service URL retrieved", "url", serviceURL)
	activity.RecordHeartbeat(ctx, "Service URL retrieved")

	// Step 5: Verify deployment health
	logger.Info("[5/5] Verifying deployment health")
	time.Sleep(2 * time.Second) // Give pods time to stabilize
	
	logger.Info("Kubernetes deployment completed successfully",
		"environment", req.Environment,
		"deploymentURL", serviceURL)

	return &shared.DeployToKubernetesResponse{
		Success:       true,
		DeploymentURL: serviceURL,
		Message:       fmt.Sprintf("Successfully deployed %s to %s", req.ImageTag, req.Environment),
		Timestamp:     time.Now(),
	}, nil
}

// CheckDeploymentStatus checks the status of a Kubernetes deployment
func (k *KubernetesActivities) CheckDeploymentStatus(ctx context.Context, req shared.CheckDeploymentStatusRequest) (*shared.CheckDeploymentStatusResponse, error) {
	logger := activity.GetLogger(ctx)
	
	logger.Info("Checking deployment status",
		"environment", req.Environment,
		"namespace", k.getNamespace(req.Environment))

	// Simulate status check
	time.Sleep(1 * time.Second)

	// In a real implementation, this would query the Kubernetes API
	return &shared.CheckDeploymentStatusResponse{
		Ready:        true,
		Replicas:     3,
		ReadyReplicas: 3,
		Message:      "All pods are running and ready",
	}, nil
}


// GetServiceURL retrieves the service URL for a deployment
func (k *KubernetesActivities) GetServiceURL(ctx context.Context, req shared.GetServiceURLRequest) (*shared.GetServiceURLResponse, error) {
	logger := activity.GetLogger(ctx)
	
	logger.Info("Getting service URL",
		"environment", req.Environment,
		"serviceName", req.ServiceName)

	// Simulate service lookup
	time.Sleep(500 * time.Millisecond)

	// Generate URL based on environment
	var serviceURL string
	if req.Environment == "staging" {
		serviceURL = fmt.Sprintf("http://staging.%s.local:8080", req.ServiceName)
	} else {
		serviceURL = fmt.Sprintf("https://%s.production.com", req.ServiceName)
	}

	return &shared.GetServiceURLResponse{
		URL:     serviceURL,
		Ready:   true,
		Message: "Service is accessible",
	}, nil
}

// Helper method to get namespace based on environment
func (k *KubernetesActivities) getNamespace(environment string) string {
	if k.Namespace != "" {
		return k.Namespace
	}
	
	switch environment {
	case "staging":
		return "staging"
	case "production":
		return "production"
	default:
		return "default"
	}
}

// createDeployment creates a new Kubernetes deployment
func (k *KubernetesActivities) createDeployment(ctx context.Context, name, image, namespace string) error {
	logger := activity.GetLogger(ctx)
	
	// Create deployment YAML
	deploymentYAML := fmt.Sprintf(`
apiVersion: apps/v1
kind: Deployment
metadata:
  name: %s
  namespace: %s
spec:
  replicas: 3
  selector:
    matchLabels:
      app: %s
  template:
    metadata:
      labels:
        app: %s
    spec:
      containers:
      - name: %s
        image: %s
        imagePullPolicy: Always
        ports:
        - containerPort: 8080
`, name, namespace, name, name, name, image)

	// Apply the deployment
	cmd := exec.Command("kubectl", "apply", "-f", "-")
	cmd.Stdin = strings.NewReader(deploymentYAML)
	
	var out bytes.Buffer
	var errOut bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errOut
	
	if err := cmd.Run(); err != nil {
		logger.Error("Failed to create deployment", "error", err, "stderr", errOut.String())
		return fmt.Errorf("failed to create deployment: %s", errOut.String())
	}
	
	logger.Info("Deployment created", "output", out.String())
	return nil
}

// ensureService ensures a Kubernetes service exists for the deployment
func (k *KubernetesActivities) ensureService(ctx context.Context, name, namespace string) error {
	logger := activity.GetLogger(ctx)
	
	// Check if service exists
	checkCmd := exec.Command("kubectl", "get", "service", name, "-n", namespace)
	if err := checkCmd.Run(); err == nil {
		logger.Info("Service already exists")
		return nil
	}
	
	// Create service YAML
	serviceYAML := fmt.Sprintf(`
apiVersion: v1
kind: Service
metadata:
  name: %s
  namespace: %s
spec:
  selector:
    app: %s
  ports:
  - port: 80
    targetPort: 8080
  type: LoadBalancer
`, name, namespace, name)

	// Apply the service
	cmd := exec.Command("kubectl", "apply", "-f", "-")
	cmd.Stdin = strings.NewReader(serviceYAML)
	
	var out bytes.Buffer
	var errOut bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errOut
	
	if err := cmd.Run(); err != nil {
		logger.Error("Failed to create service", "error", err, "stderr", errOut.String())
		return fmt.Errorf("failed to create service: %s", errOut.String())
	}
	
	logger.Info("Service created", "output", out.String())
	return nil
}

// getActualServiceURL gets the actual URL for the Kubernetes service
func (k *KubernetesActivities) getActualServiceURL(ctx context.Context, name, namespace string) (string, error) {
	logger := activity.GetLogger(ctx)
	
	// Try to get external IP/hostname from LoadBalancer service
	cmd := exec.Command("kubectl", "get", "service", name, "-n", namespace, 
		"-o", "jsonpath={.status.loadBalancer.ingress[0].hostname}{.status.loadBalancer.ingress[0].ip}")
	
	var out bytes.Buffer
	var errOut bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errOut
	
	if err := cmd.Run(); err != nil {
		logger.Warn("Failed to get LoadBalancer URL, trying NodePort", "error", err, "stderr", errOut.String())
		// If LoadBalancer is not available, try NodePort
		return k.getNodePortURL(ctx, name, namespace)
	}
	
	externalAddr := strings.TrimSpace(out.String())
	logger.Info("LoadBalancer external address", "address", externalAddr, "namespace", namespace)
	
	if externalAddr == "" {
		logger.Warn("No external address found, trying NodePort")
		return k.getNodePortURL(ctx, name, namespace)
	}
	
	// Determine protocol based on environment
	protocol := "http"
	if namespace == "production" {
		protocol = "https"
	}
	
	serviceURL := fmt.Sprintf("%s://%s", protocol, externalAddr)
	logger.Info("Generated LoadBalancer service URL", "url", serviceURL)
	
	return serviceURL, nil
}

// getNodePortURL gets the NodePort URL as a fallback
func (k *KubernetesActivities) getNodePortURL(ctx context.Context, name, namespace string) (string, error) {
	logger := activity.GetLogger(ctx)
	
	// Get node IP
	nodeCmd := exec.Command("kubectl", "get", "nodes", "-o", 
		"jsonpath={.items[0].status.addresses[?(@.type=='InternalIP')].address}")
	
	var nodeOut bytes.Buffer
	var nodeErr bytes.Buffer
	nodeCmd.Stdout = &nodeOut
	nodeCmd.Stderr = &nodeErr
	
	if err := nodeCmd.Run(); err != nil {
		logger.Error("Failed to get node IP", "error", err, "stderr", nodeErr.String())
		// Return a default URL if we can't get the actual one
		if namespace == "staging" {
			return "http://staging.demo-app.local:8080", nil
		}
		return "https://demo-app.production.local", nil
	}
	
	nodeIP := strings.TrimSpace(nodeOut.String())
	
	// Get NodePort
	portCmd := exec.Command("kubectl", "get", "service", name, "-n", namespace,
		"-o", "jsonpath={.spec.ports[0].nodePort}")
	
	var portOut bytes.Buffer
	var portErr bytes.Buffer
	portCmd.Stdout = &portOut
	portCmd.Stderr = &portErr
	
	if err := portCmd.Run(); err != nil {
		logger.Error("Failed to get NodePort", "error", err, "stderr", portErr.String())
		// Return a default URL if we can't get the actual one
		if namespace == "staging" {
			return "http://staging.demo-app.local:8080", nil
		}
		return "https://demo-app.production.local", nil
	}
	
	nodePort := strings.TrimSpace(portOut.String())
	
	return fmt.Sprintf("http://%s:%s", nodeIP, nodePort), nil
}